src/monodroid/jni/strings.hh

```diff
@@ -0,0 +1,731 @@
+
+	template<size_t MaxStackSize, typename T = char>
+	class static_local_storage final : public static_local_storage_base<MaxStackSize, T>
+	{
+		using base = static_local_storage_base<MaxStackSize, T>;
+
+	public:
+		explicit static_local_storage (size_t initial_size) noexcept
+			: base (initial_size)
+		{}
+	};
+
+	template <size_t MaxStackSize, typename T> using dynamic_local_storage_base = local_storage<MaxStackSize, true, T>;
+
+	template<size_t MaxStackSize, typename T = char>
+	class dynamic_local_storage final : public dynamic_local_storage_base<MaxStackSize, T>
+	{
+		using base = dynamic_local_storage_base<MaxStackSize, T>;
+
+	public:
+		explicit dynamic_local_storage (size_t initial_size = 0) noexcept
+			: base (initial_size)
+		{}
+
+		//
+		// If `new_size` is smaller than the current size and the dynamic store is allocated, data WILL NOT be
+		// preserved.
+		//
+		// If `new_size` is bigger than the current size bigger than MaxStackSize, data will be copied to the new
+		// dynamically allocated store
+		//
+		// If `new_size` is smaller or equal to `MaxStackSize`, no changes are made unless dynamic store was allocated,
+		// in which case it will be freed
+		//
+		void resize (size_t new_size) noexcept
+		{
+			size_t old_size = base::size ();
+
+			if (new_size == old_size) {
+				return;
+			}
+
+			if (new_size <= MaxStackSize) {
+				new_size = MaxStackSize;
+				base::free_store ();
+				return;
+			}
+
+			if (new_size < old_size) {
+				base::free_store ();
+				base::init_store (new_size);
+				return;
+			}
+
+			T* old_allocated_store = base::get_allocated_store ();
+			base::init_store (new_size);
+
+			T* new_allocated_store = base::get_allocated_store ();
+			if (old_allocated_store != nullptr) {
+				std::memcpy (new_allocated_store, old_allocated_store, old_size);
+				delete[] old_allocated_store;
+				return;
+			}
+
+			std::memcpy (new_allocated_store, base::get_local_store ().data (), MaxStackSize);
+		}
+	};
+
+	// This class mostly performs NO BOUNDS CHECKING - child classes MUST implement that!
+	template<size_t MaxStackSize, typename TStorage, typename TChar = char>
+	class string_base
+	{
+	protected:
+		static constexpr TChar NUL = '\0';
+		static constexpr TChar ZERO = '0';
+
+	public:
+		explicit string_base (size_t initial_size = 0)
+			: buffer (initial_size)
+		{
+			// Currently we care only about `char`, maybe in the future we'll add support for `wchar` (if needed)
+			static_assert (std::is_same_v<TChar, char>, "TChar must be an 8-bit character type");
+
+			clear ();
+		}
+
+		explicit string_base (const string_segment &token)
+			: string_base (token.initialized () ? token.length () : 0)
+		{
+			if (token.initialized ())
+				assign (token.start (), token.length ());
+		}
+
+		force_inline size_t index () const noexcept
+		{
+			return idx;
+		}
+
+		force_inline size_t length () const noexcept
```

Comment:
Why have two member functions with the same semantics, `index()` and `length()`?